content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

```diff
@@ -671,681 +880,158 @@ subprogram cannot be called if :ada:`Index` is out of the boundaries of data:
-       begin
-          if D = North and then Y < Max then
-             Y := Y + 1;
-          elsif D = South and then Y > Min  then
-             Y := Y - 1;
-          elsif D = West and then X > Min then
-             X := X - 1;
-          elsif D = East and then X < Max then
-             X := X + 1;
-          end if;
-       end Move;
-
-    end Directions;
-
-:code-config:`accumulate_code=False`
-
-:code-config:`reset_accumulator=True`
-
-RavenSPARK
-----------
-
-.. todo::
-
-    Complete section!
-
-
-Physical Dimension Checking
----------------------------
-
-.. todo::
+   type Move_Result is (Full_Speed, Slow_Down, Keep_Going, Stop);
+
+   procedure Move
+      (Train        : in  Train_Id;
+       New_Position : in  Train_Position;
+       Result       : out Move_Result)
+   with
+      Pre  => Valid_Id (Train) and
+              Valid_Move (Trains (Train), New_Position) and
+              At_Most_One_Train_Per_Track and
+              Safe_Signaling,
+      Post => At_Most_One_Train_Per_Track and
+              Safe_Signaling;
+
+   function At_Most_One_Train_Per_Track return Boolean;
+
+   function Safe_Signaling return Boolean;
+
+The preconditions specify that, given a safe initial state and a valid
+move, the result of the call will also be a safe state: there will be at
+most one train per track section and the track signaling system will not
+allow any unsafe movements.
+
+
+Final Comments
+--------------
+
+Make sure you understand that :program:`gnatprove` does not attempt to 
+prove the program correct as a whole. It attempts to prove 
+language-defined and user-defined assertions about parts of the program, 
+especially individual routines and calls to those routines. Furthermore, 
+:program:`gnatprove` proves the routines correct only to the extent that 
+the user-defined assertions correctly and sufficiently describe and 
+constrain the implementation of the corresponding routines. 
+
+Although we are not proving whole program correctness, as you will have 
+seen |mdash| and done |mdash| we can prove properties than make our 
+software far more robust and bug-free than is possible otherwise. But in 
+addition, consider what proving the unit-level requirements for your 
+procedures and functions would do for the cost of unit testing and 
+system integration. The tests would pass the first time. 
+
+However, within the scope of what SPARK can do, not everything can be 
+proven. In some cases that is because the software behavior is not 
+amenable to expression as boolean conditions (for example, a mouse 
+driver). In other cases the source code is beyond the capabilities of 
+the analyzers that actually do the mathematical proof. In these cases 
+the combination of proof and actual test is appropriate, and still less 
+expensive that testing alone. 
+
+There is, of course, much more to be said about what can be done with 
+SPARK and :program:`gnatprove`. Those topics are reserved for the 
+advanced SPARK course. 
```

Comment:
I think we should link to the existing _Intro to SPARK_ course instead. This would be the appropriate syntax:
```
:doc:`Introduction to SPARK <courses/intro-to-spark/index>` course.
```